graphql_compiler/compiler/ir_lowering_match.py

```diff
@@ -451,127 +455,215 @@ def remove_backtrack_blocks_from_fold(folded_ir_blocks):
             match_traversals = match_query.match_traversals
             output_block = match_query.output_block
             folds = match_query.folds
-            current_locations = set()
-            current_non_optional_locations = set()
 
-            for traversal in match_traversals:
-                for step in traversal:
-                    if step.as_block is not None:
-                        location_name, _ = step.as_block.location.get_location_name()
-                        current_locations.add(location_name)
-                        if isinstance(step.root_block, Traverse) and not step.root_block.optional:
-                            current_non_optional_locations.add(location_name)
+            present_locations_tuple = _get_present_locations_from_match_traversals(match_traversals)
+            present_locations, present_non_optional_locations = present_locations_tuple
 
             new_output_fields = {}
             for output_name, expression in six.iteritems(output_block.fields):
-                # If @fold is allowed within @optional, this should include FoldedOutputContextField
                 if isinstance(expression, OutputContextField):
                     location_name, _ = expression.location.get_location_name()
-                    if location_name not in current_locations:
+                    if location_name not in present_locations:
+                        raise AssertionError(u'Non-optional output location {} was not found in '
+                                             u'present_locations: {}'
+                                             .format(expression.location, present_locations))
+                    new_output_fields[output_name] = expression
+                elif isinstance(expression, FoldedOutputContextField):
+                    base_location = expression.fold_scope_location.base_location
+                    location_name, _ = base_location.get_location_name()
+                    if location_name not in present_locations:
                         raise AssertionError(u'Non-optional output location {} was not found in '
-                                             u'current_locations: {}'
-                                             .format(expression.location, current_locations))
+                                             u'present_locations: {}'
+                                             .format(expression.location, present_locations))
                     new_output_fields[output_name] = expression
                 elif isinstance(expression, TernaryConditional):
                     location_name, _ = expression.if_true.location.get_location_name()
-                    if location_name in current_locations:
-                        if location_name in current_non_optional_locations:
+                    if location_name in present_locations:
+                        if location_name in present_non_optional_locations:
```

Comment:
If `present_non_optional_locations` is guaranteed to be a subset of `present_locations` as I asked above, then the `if` on line 659 isn't necessary, no?